### PR TITLE
decay types in wrapping code, not in enable if

### PR DESCRIPTION
**Pull request by jimhester**
This simplifies the enable if logic and should help improve error
messages

### CODE REVIEW

**Comment by jimhester** [comment]
@bkietz if you could look this over, I think it is what you had in mind.

File: cpp11test/src/test-as.cpp [hunk 1]
```diff
@@ -285,7 +285,7 @@ context("as_cpp-C++") {
     expect_true(x2[1] == 2.);
     expect_true(x2[2] == 3.);
 
-    auto x3 = cpp11::as_cpp<const cpp11::doubles&>(r);
+    auto x3 = cpp11::as_cpp<cpp11::decay_t<const cpp11::doubles&>>(r);
```
Comment (bkietz) [inline cpp11test/src/test-as.cpp:288]: I'd just delete these test cases; no human should write `as_cpp<non_decayed>`

File: cpp11test/src/test-as.cpp [hunk 2]
```diff
@@ -476,9 +476,10 @@ context("as_cpp-C++") {
     UNPROTECT(1);
   }
 
-  test_that("as_sexp(r_vector<cpp11::r_string>)") {
-    SEXP s1 =
-        PROTECT(cpp11::as_sexp(std::vector<cpp11::r_string>({"foo", "bar", "baz"})));
+  test_that("as_sexp(cpp11::strings)") {
+    cpp11::writable::strings s({"foo", "bar", "baz"});
+    cpp11::strings s0(s);
+    SEXP s1 = PROTECT(cpp11::as_sexp(s0));
```
Comment (bkietz) [inline cpp11test/src/test-as.cpp:482]: this isn't transparent to me; did compilation break or is it just a preferred spelling of this test or ...?

File: inst/include/cpp11/as.hpp [hunk 3]
```diff
@@ -174,14 +173,6 @@ enable_if_std_string<T, T> as_cpp(SEXP from) {
   return {as_cpp<const char*>(from)};
 }
```
Comment (bkietz) [inline inst/include/cpp11/as.hpp:175]:
```suggestion
/// If your compilation fails here you are calling as_cpp<T> with a T
/// which is not decayed; consider rewriting or wrapping T into cpp11::decay_t<T>.
template <typename T>
enable_if_t<!std::is_same<decay_t<T>, T>::value, T> as_cpp(SEXP from) = delete;
```

File: cpp11test/src/test-as.cpp (hunk 1 above)
Comment (jimhester) [reply cpp11test/src/test-as.cpp:288]: yeah, I guess there is little point in keeping them

File: cpp11test/src/test-as.cpp (hunk 2 above)
Comment (jimhester) [reply cpp11test/src/test-as.cpp:482]: Yeah this particular case `as_sexp(cpp11::string)` is something that occurs in [vroom](https://github.com/r-lib/vroom) but we didn't use it in the cpp11test code at all, so I added a test case for it specifically.

**Comment by bkietz** [comment]
@jimhester re `unmove`: this *will* be necessary if we want to support registering functions with parameters which are mutable references. I think it'd be better not to support these however since the connotation of a mutable reference parameter is a modifiable object from an enclosing scope, whereas we'll always be providing a temporary (and unmove just allows us to wrap that temporary in an ostensibly mutable reference). Specifically, if a user registers

```c++
void append_1(std::vector<std::string>& char_vec) {
  char_vec.push_back("one");
}
```

they might reasonably expect that to mutate a character vector in R the same way calling it from other C++ code would.

File: cpp11test/src/test-as.cpp (hunk 2 above)
Comment (bkietz) [reply cpp11test/src/test-as.cpp:459]: alright, but let's keep the original one as well
```suggestion
  test_that("as_sexp(cpp11::strings)") {
    cpp11::writable::strings s({"foo", "bar", "baz"});
    cpp11::strings s0(s);
    SEXP s1 = PROTECT(cpp11::as_sexp(s0));
    
    SEXP s2 = PROTECT(cpp11::as_sexp(std::vector<cpp11::r_string>({"foo", "bar", "baz"})));
```

**Comment by bkietz** [comment]
@jimhester looks like there are a few remaining test cases which call `as_cpp<non decayed>` https://github.com/r-lib/cpp11/pull/75/checks?check_run_id=950884317#step:7:122

**Comment by bkietz** [comment]
@jimhester these changes seem to have been merged but the PR is still open; I think you made the changes to r-lib/master locally and pushed?

**Comment by jimhester** [comment]
yeah, I'll fix it soon

**Pull request merged**